### PR TITLE
ZEN-34366: allow -replacement and -addition templates for non-ZPL dev…

### PR DESCRIPTION
**Pull request by StevePC**
…ices and components.  This will require an update to ZPL's DeviceBase.getRRDTemplates() which assumes replaced templates have not already been removed from the list

### CODE REVIEW

**Comment by webhookbotuser** [comment]
Can one of the admins verify this patch?

**Comment by webhookbotuser** [comment]
Can one of the admins verify this patch?

**Comment by webhookbotuser** [comment]
Can one of the admins verify this patch?

File: Products/ZenModel/Device.py [hunk 1]
```diff
@@ -400,12 +400,36 @@ def getRRDTemplates(self):
...
+            if addition and addition not in templates:
+                templates.append(addition)
+                #self.setZenProperty(
+                #    'zDeviceTemplates',
+                #    self.zDeviceTemplates + [addition.titleOrId()])
+
             if template:
-                result.append(template)
-        return result
+                templates.append(template)
```
Comment (zensaiello) [inline Products/ZenModel/Device.py:431]: This code block logic is replaced by 418-419, 'Only add if -replacement named template does not exist' ?

**Review by zensaiello — CHANGES_REQUESTED** [review]

File: Products/ZenModel/IpInterface.py [hunk 2]
```diff
@@ -574,9 +574,23 @@ def getRRDTemplates(self):
...
             template = self.getRRDTemplateByName(name)
-            if template:
-                return [template]
+            replacement = self.getRRDTemplateByName(
+                '{}-replacement'.format(templateName))
+
+            if replacement and replacement not in templates:
+                templates.append(replacement)
+            else:
+                templates.append(template)
```
Comment (Deer-WarLord) [inline Products/ZenModel/IpInterface.py:585]: Don't we need to check whether the template is none?

File: Products/ZenModel/MetricMixin.py [hunk 3]
```diff
@@ -139,10 +139,19 @@ def snmpIgnore(self):
...
-        if not default:
-            return []
-        return [default]
+        templates = []
+        defaultName = self.getRRDTemplateName()
+        replacement = self.getRRDTemplateByName('{}-replacement'.format(defaultName))
+        if replacement:
+            templates.append(replacement)
+        else:
+            default = self.getRRDTemplateByName(defaultName)
```
Comment (Deer-WarLord) [inline Products/ZenModel/MetricMixin.py:148]: Why do we get **default** only if the replacement is none? Shouldn't the approach be the same as for IpInterface.py?

**Review by Deer-WarLord — CHANGES_REQUESTED** [review]

File: Products/ZenModel/Device.py [hunk 4]
```diff
@@ -400,12 +400,36 @@ def getRRDTemplates(self):
...
+
+            template = self.getRRDTemplateByName(templateName)
+            replacement = self.getRRDTemplateByName(
+                '{}-replacement'.format(templateName))
+
+            if replacement and replacement not in templates:
+                templates.append(replacement)
+                #self.setZenProperty(
+                #    'zDeviceTemplates',
+                #    self.zDeviceTemplates + [replacement.titleOrId()])
```
Comment (jpeacock-zenoss) [inline Products/ZenModel/Device.py:417]: Commented out code should be removed.  It's not doing anything and doesn't help explain (to me) the surrounding code.

File: Products/ZenModel/Device.py [hunk 5]
```diff
@@ -400,12 +400,36 @@ def getRRDTemplates(self):
...
+                templates.append(template)
+
+            addition = self.getRRDTemplateByName(
+                '{}-addition'.format(templateName))
+
+            if addition and addition not in templates:
+                templates.append(addition)
+                #self.setZenProperty(
+                #    'zDeviceTemplates',
+                #    self.zDeviceTemplates + [addition.titleOrId()])
```
Comment (jpeacock-zenoss) [inline Products/ZenModel/Device.py:428]: Same thing about commented out code.

**Review by jpeacock-zenoss — CHANGES_REQUESTED** [review]

**Review by zensaiello — APPROVED** [review]

**Review by Deer-WarLord — APPROVED** [review]

**Review by StevePC — COMMENTED** [review]
retest this please

**Review by jpeacock-zenoss — APPROVED** [review]

**Pull request merged**